certstore_linux.go

```diff
@@ -1,14 +1,336 @@
+		case crypto.SHA512:
+			mechanism = C.CKM_SHA512_RSA_PKCS
+		*/
+		default:
+			mechanism = C.CKM_RSA_PKCS
+		}
+	case C.ecKey:
+		switch hash {
+		/*
+		case crypto.SHA1:
+			mechanism = C.CKM_ECDSA_SHA1
+		*/
+		default:
+			mechanism = C.CKM_ECDSA
+		}
+	default:
+		return nil, fmt.Errorf("Unknown key type: %d", int(key.keyType))
+	}
+	ret := C.PK11_SignWithMechanism(key, mechanism, nil, sd, hashed)
+	if ret != 0 {
+		return nil, errors.New("could not sign")
+	}
+	var sig = C.GoBytes(unsafe.Pointer(sd.data), C.int(sd.len))
+	if key.keyType == C.ecKey {
+		if len(sig)%2 != 0 {
+			return nil, errors.New("bad ecdsa signature")
+		}
+		type ecdsaSignature struct {
+			R, S *big.Int
+		}
+		r := new(big.Int).SetBytes(sig[:len(sig)/2])
+		s := new(big.Int).SetBytes(sig[len(sig)/2:])
+		asn, err := asn1.Marshal(ecdsaSignature{r, s})
+		if err != nil {
+			return nil, err
+		}
+		sig = asn
+	}
+	return sig, nil
+}
+
+func (nssStore) Import(data []byte, password string) error {
+	unicodePassword, err := bmpString(password)
+	if unicodePassword == nil {
+		return err
+	}
+	var pass = C.SECITEM_AllocItem(nil, nil, C.uint(len(unicodePassword)))
+	if pass == nil {
+		return errors.New("SECITEM_AllocItem failed")
+	}
+	C.memcpy(unsafe.Pointer(pass.data), unsafe.Pointer(&unicodePassword[0]), C.size_t(len(unicodePassword)))
+	var p12 = C.SEC_PKCS12DecoderStart(pass, nil, nil, nil, nil, nil, nil, nil)
+	var decoded = C.SEC_PKCS12DecoderUpdate(p12, (*C.uchar)(unsafe.Pointer(&data[0])), C.size_t(len(data)))
+	if decoded != 0 {
+		return fmt.Errorf("Error %d, P12 decoding failed...\n", int(C.PR_GetError()))
+	}
+	var authenticated = C.SEC_PKCS12DecoderVerify(p12)
+	if authenticated != 0 {
+		return fmt.Errorf("Error %d, P12 authentication failed...\n", int(C.PR_GetError()))
+	}
+	var validated = C.SEC_PKCS12DecoderValidateBags(p12, (*[0]byte)(C.P12U_NicknameCollisionCallback))
+	if validated != 0 {
+		return fmt.Errorf("Error %d, P12 validation failed...\n", int(C.PR_GetError()))
+	}
+	var imported = C.SEC_PKCS12DecoderImportBags(p12)
+	if imported != 0 {
+		return fmt.Errorf("Error %d, P12 import failed...\n", int(C.PR_GetError()))
+	}
+	return nil
+}
+
+func (nssStore) Close() {
 }
 
-// Implement this function, just to silence other compiler errors.
 func openStore() (Store, error) {
-	return nil, errors.New("certstore only works on macOS and Windows")
+	var homeDir = os.Getenv("HOME")
+
+	if homeDir == "" {
+		user, err := user.Current()
+
+		if (err == nil) {
+			homeDir = user.HomeDir
+		}
+	}
+
+	if (homeDir == "") {
+		return nil, errors.New("Unable to locate user's home directory")
+	}
+
+	var nssdb = path.Join(homeDir, ".pki", "nssdb")
+	if _, err := os.Stat(nssdb); os.IsNotExist(err) {
+		return nil, fmt.Errorf("NSS database not found at %s\n", nssdb)
+	}
+	nssdbUrlC := C.CString(fmt.Sprintf("sql:/%s/", nssdb))
+	defer C.free(unsafe.Pointer(nssdbUrlC))
+	//fmt.Printf("Opening: %s\n", nssdbUrl)
+	ok := C.NSS_InitReadWrite(nssdbUrlC)
```

Comment:
I wonder if we should be using `NSS_InitContext` instead to avoid messing with other libraries that might be using NSS.

Reply:
Yes, I think we should... https://wiki.mozilla.org/NSS_Library_Init

Reply:
OK, I've switched the code to use `NSS_InitContext`. It's maybe not optimal, but it passes tests :woman_shrugging: